### PR TITLE
Allow assembly inside pil.

### PR DESCRIPTION
**Pull request by chriseth**
Depends on #291

### CODE REVIEW

File: parser/src/display.rs [hunk 1]
```diff
@@ -76,6 +76,14 @@ impl<T: Display> Display for Statement<T> {
             Statement::FunctionCall(_, name, args) => {
                 write!(f, "{name}({});", format_expressions(args))
             }
+            Statement::ASMBlock(_, statements) => {
+                writeln!(f, "assembly {{")?;
+                for _s in statements {
+                    // TODO display for asm statements
```
Comment (chriseth) [inline parser/src/display.rs:82]: We don't have it at all yet.

File: compiler/tests/pil.rs [hunk 2]
```diff
@@ -69,3 +69,17 @@ fn test_pair_lookup() {
...
+#[test]
+fn test_simple_sum_asm_pil() {
+    verify_pil(
+        "simple_sum_asm.pil",
+        Some(|q| match q {
+            "\"input\", 0" => Some(13.into()),
+            "\"input\", 1" => Some(2.into()),
+            "\"input\", 2" => Some(11.into()),
+            "\"input\", 3" => Some(2.into()),
+            _ => Some(7.into()),
```
Comment (leonardoalt) [inline compiler/tests/pil.rs:82]: why 7?

File: pil_analyzer/src/pil_analyzer.rs [hunk 3]
```diff
@@ -454,6 +460,13 @@ impl<T: FieldElement> PILContext<T> {
         assert!(is_new);
     }
 
+    fn handle_assembly(&mut self, _source: SourceRef, asm_statements: Vec<ASMStatement<T>>) {
+        let statements = pilgen::asm_to_pil(asm_statements.into_iter());
```
Comment (leonardoalt) [inline pil_analyzer/src/pil_analyzer.rs:464]: So the assembly blocks are always self contained?

File: compiler/tests/pil.rs (hunk 2 above)
Comment (chriseth) [reply compiler/tests/pil.rs:82]: Just a random number.

File: pil_analyzer/src/pil_analyzer.rs (hunk 3 above)
Comment (chriseth) [reply pil_analyzer/src/pil_analyzer.rs:464]: What do you mean? They cannot have include statements?

**Comment by chriseth** [comment]
Rebased.

File: pil_analyzer/src/pil_analyzer.rs (hunk 3 above)
Comment (leonardoalt) [reply pil_analyzer/src/pil_analyzer.rs:464]: yea that they cannot depend on this outside of it

Comment (chriseth) [reply pil_analyzer/src/pil_analyzer.rs:464]: What do you mean by depend? Both assembly and pil inside and outside of assembly can freely reference names in all those areas.

Comment (leonardoalt) [reply pil_analyzer/src/pil_analyzer.rs:464]: Right yea, maybe we should clarify the scoping rules in the future.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**